File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/PythonManifest.kt
package dev.pacmon.jetbrains.core

/** Static Python dependency parsers. Pacmon never invokes Python, pip, Poetry, uv, or a build backend. */
object PythonManifestAdapter : ManifestAdapter {
    override val kind = ManifestKind.PYTHON
    override val fileNames = listOf("pyproject.toml", "requirements.txt")
    override val discoveryGlobs = listOf("**/pyproject.toml", "**/requirements*.txt", "**/requirements/**/*.txt")
    override val notesRelativePath = ".pacmon/python/DEPENDENCY-NOTES.md"

    override fun matchesPath(path: String): Boolean {
        val normalized = path.replace('\\', '/')
        val basename = normalized.substringAfterLast('/')
        return basename == "pyproject.toml" ||
            (basename.startsWith("requirements") && basename.endsWith(".txt")) ||
            (basename.endsWith(".txt") && normalized.split('/').contains("requirements"))
    }

    override fun extractDependencies(text: String): List<DependencyEntry> = extractDependencies(text, "pyproject.toml")

    override fun extractDependencies(text: String, path: String): List<DependencyEntry> =
        if (path.replace('\\', '/').substringAfterLast('/') == "pyproject.toml") extractPyproject(text)
        else extractRequirements(text, requirementsScope(path))

    override fun normalizeNoteKey(raw: String): String = normalizePackageName(ManifestRegistry.stripName(raw))

    fun normalizePackageName(value: String): String = value.trim().lowercase().replace(Regex("[._-]+"), "-")

    private data class Line(val text: String, val offset: Int)
    private data class Assignment(
        val table: List<String>,
        val key: String,
        val keyRange: SourceRange,
        val valueText: String,
        val valueOffset: Int,
    )

    private fun lines(text: String): List<Line> = buildList {
        var offset = 0
        while (offset <= text.length) {
            val newline = text.indexOf('\n', offset)
            val rawEnd = if (newline < 0) text.length else newline
            val end = if (rawEnd > offset && text[rawEnd - 1] == '\r') rawEnd - 1 else rawEnd
            add(Line(text.substring(offset, end), offset))
            if (newline < 0) break
            offset = newline + 1
        }
    }

    private fun commentStart(value: String): Int {
        var quote: Char? = null
        var escaped = false
        value.forEachIndexed { index, char ->
            if (quote == '"' && escaped) escaped = false
            else if (quote == '"' && char == '\\') escaped = true
            else if (quote != null) {
                if (char == quote) quote = null
            } else if (char == '"' || char == '\'') quote = char
            else if (char == '#') return index
        }
        return value.length
    }

    private fun splitDottedKey(raw: String): List<String>? {
        val parts = mutableListOf<String>()
        var start = 0
        var quote: Char? = null
        var escaped = false
        fun push(end: Int): Boolean {
            val token = raw.substring(start, end).trim()
            if (token.isEmpty()) return false
            parts += if (token.length >= 2 && token.first() in setOf('"', '\'') && token.last() == token.first()) {
                token.substring(1, token.length - 1)
            } else token
            return true
        }
        raw.forEachIndexed { index, char ->
            if (quote == '"' && escaped) escaped = false
            else if (quote == '"' && char == '\\') escaped = true
            else if (quote != null) {
                if (char == quote) quote = null
            } else if (char == '"' || char == '\'') quote = char
            else if (char == '.') {
                if (!push(index)) return null
                start = index + 1
            }
        }
        return if (quote == null && push(raw.length)) parts else null
    }

    private fun unquotedEquals(raw: String): Int {
        var quote: Char? = null
        var escaped = false
        raw.forEachIndexed { index, char ->
            if (quote == '"' && escaped) escaped = false
            else if (quote == '"' && char == '\\') escaped = true
            else if (quote != null) {
                if (char == quote) quote = null
            } else if (char == '"' || char == '\'') quote = char
            else if (char == '=') return index
        }
        return -1
    }

    private fun nestingDelta(raw: String): Int {
        var delta = 0
        var quote: Char? = null
        var escaped = false
        for (char in raw) {
            if (quote == '"' && escaped) escaped = false
            else if (quote == '"' && char == '\\') escaped = true
            else if (quote != null) {
                if (char == quote) quote = null
            } else if (char == '"' || char == '\'') quote = char
            else if (char == '[' || char == '{') delta++
            else if (char == ']' || char == '}') delta--
        }
        return delta
    }

    private fun assignments(text: String): List<Assignment> {
        val sourceLines = lines(text)
        val out = mutableListOf<Assignment>()
        var table = emptyList<String>()
        var index = 0
        while (index < sourceLines.size) {
            val line = sourceLines[index]
            val content = line.text.substring(0, commentStart(line.text))
            val trimmed = content.trim()
            if (trimmed.isEmpty()) {
                index++
                continue
            }
            if (trimmed.startsWith('[')) {
                table = if (!trimmed.startsWith("[[") && trimmed.endsWith(']')) {
                    splitDottedKey(trimmed.substring(1, trimmed.length - 1)) ?: emptyList()
                } else emptyList()
                index++
                continue
            }
            val equals = unquotedEquals(content)
            if (equals < 0) {
                index++
                continue
            }
            val rawKey = content.substring(0, equals)
            val keyToken = rawKey.trim()
            val keyParts = splitDottedKey(keyToken)
            if (keyParts.isNullOrEmpty()) {
                index++
                continue
            }
            val keyLeading = rawKey.length - rawKey.trimStart().length
            val quoted = keyToken.length >= 2 && keyToken.first() in setOf('"', '\'') && keyToken.last() == keyToken.first()
            val keyRange = SourceRange(line.offset + keyLeading + if (quoted) 1 else 0, keyToken.length - if (quoted) 2 else 0)
            var valueText = content.substring(equals + 1)
            val valueOffset = line.offset + equals + 1
            var depth = nestingDelta(valueText)
            while (depth > 0 && index + 1 < sourceLines.size) {
                val next = sourceLines[++index]
                val nextContent = next.text.substring(0, commentStart(next.text))
                valueText += " ".repeat((next.offset - (valueOffset + valueText.length)).coerceAtLeast(0)) + nextContent
                depth += nestingDelta(nextContent)
            }
            out += Assignment(table, keyParts.joinToString("."), keyRange, valueText, valueOffset)
            index++
        }
        return out
    }

    private data class TomlString(val raw: String, val contentOffset: Int, val range: SourceRange)

    private fun arrayStrings(assignment: Assignment): List<TomlString> = buildList {
        val raw = assignment.valueText
        var inlineTableDepth = 0
        var index = 0
        while (index < raw.length) {
            val quote = raw[index]
            if (quote == '{') {
                inlineTableDepth++
                index++
                continue
            }
            if (quote == '}') {
                inlineTableDepth = (inlineTableDepth - 1).coerceAtLeast(0)
                index++
                continue
            }
            if (quote !in setOf('"', '\'')) {
                index++
                continue
            }
            if (raw.substring(index).startsWith(quote.toString().repeat(3))) {
                val end = raw.indexOf(quote.toString().repeat(3), index + 3)
                if (end < 0) break
                index = end + 3
                continue
            }
            val start = index++
            var escaped = false
            while (index < raw.length) {
                val char = raw[index]
                if (quote == '"' && escaped) escaped = false
                else if (quote == '"' && char == '\\') escaped = true
                else if (char == quote) break
                index++
            }
            if (index >= raw.length) break
            if (inlineTableDepth == 0) {
                add(TomlString(
                    raw.substring(start + 1, index),
                    assignment.valueOffset + start + 1,
                    SourceRange(assignment.valueOffset + start, index - start + 1),
                ))
            }
            index++
        }
    }

    private data class RequirementName(val name: String, val start: Int, val length: Int)
    private val requirementNamePattern = Regex("^\\s*([A-Za-z0-9](?:[A-Za-z0-9._-]*[A-Za-z0-9])?)")

    private fun requirementName(raw: String): RequirementName? {
        val match = requirementNamePattern.find(raw) ?: return null
        val name = match.groupValues[1]
        val end = match.range.last + 1
        val next = raw.getOrNull(end)
        if (next != null && !next.isWhitespace() && next !in "[(<>=!~@;") return null
        val start = match.value.length - name.length
        return RequirementName(name, start, name.length)
    }

    private fun requirementEntry(value: TomlString, scope: String): DependencyEntry? {
        val parsed = requirementName(value.raw) ?: return null
        val range = SourceRange(value.contentOffset + parsed.start, parsed.length)
        return dependencyEntry(normalizePackageName(parsed.name), scope, range, listOf(range), parsed.name, value.range)
    }

    private fun tableEquals(table: List<String>, vararg wanted: String): Boolean = table == wanted.toList()

    fun extractPyproject(text: String): List<DependencyEntry> = buildList {
        for (assignment in assignments(text)) {
            val arrayScope = when {
                tableEquals(assignment.table, "project") && assignment.key == "dependencies" -> "project"
                tableEquals(assignment.table, "project", "optional-dependencies") -> "extra:${assignment.key}"
                tableEquals(assignment.table, "dependency-groups") -> "group:${assignment.key}"
                tableEquals(assignment.table, "build-system") && assignment.key == "requires" -> "build-system"
                tableEquals(assignment.table, "tool", "uv") && assignment.key == "dev-dependencies" -> "uv:dev"
                else -> null
            }
            if (arrayScope != null) {
                arrayStrings(assignment).mapNotNullTo(this) { requirementEntry(it, arrayScope) }
                continue
            }
            val poetryScope = when {
                tableEquals(assignment.table, "tool", "poetry", "dependencies") -> "poetry:main"
                tableEquals(assignment.table, "tool", "poetry", "dev-dependencies") -> "poetry:dev"
                assignment.table.size == 5 && assignment.table[0] == "tool" && assignment.table[1] == "poetry" &&
                    assignment.table[2] == "group" && assignment.table[4] == "dependencies" ->
                    "poetry:group:${assignment.table[3]}"
                else -> null
            }
            if (poetryScope != null && !assignment.key.equals("python", ignoreCase = true)) {
                add(dependencyEntry(
                    normalizePackageName(assignment.key),
                    poetryScope,
                    assignment.keyRange,
                    listOf(assignment.keyRange),
                    assignment.key,
                    assignment.keyRange,
                ))
            }
        }
    }

    private data class LogicalLine(val value: String, val positions: List<Int>)

    private fun logicalLines(text: String): List<LogicalLine> = buildList {
        var value = StringBuilder()
        var positions = mutableListOf<Int>()
        for (line in lines(text)) {
            val continued = Regex("\\\\\\s*$").containsMatchIn(line.text)
            val part = if (continued) line.text.replace(Regex("\\\\\\s*$"), "") else line.text
            part.forEachIndexed { index, char ->
                value.append(char)
                positions += line.offset + index
            }
            if (continued) {
                value.append(' ')
                positions += line.offset + line.text.length
            } else {
                add(LogicalLine(value.toString(), positions))
                value = StringBuilder()
                positions = mutableListOf()
            }
        }
        if (value.isNotEmpty()) add(LogicalLine(value.toString(), positions))
    }

    private fun requirementCommentStart(value: String): Int = value.indices.firstOrNull {
        value[it] == '#' && (it == 0 || value[it - 1].isWhitespace())
    } ?: value.length

    fun extractRequirements(text: String, scope: String = "requirements.txt"): List<DependencyEntry> = buildList {
        for (logical in logicalLines(text)) {
            val raw = logical.value.substring(0, requirementCommentStart(logical.value))
            var candidateStart = raw.length - raw.trimStart().length
            var candidate = raw.trimStart()
            if (candidate.isEmpty()) continue
            if (Regex("^(?:-r|--requirement|-c|--constraint)(?:\\s|=)").containsMatchIn(candidate) ||
                candidate.startsWith("--")) continue
            val editable = Regex("^(?:-e|--editable)(?:\\s+|=)").find(candidate)
            if (editable != null) {
                candidateStart += editable.value.length
                candidate = candidate.substring(editable.value.length)
            } else if (candidate.startsWith('-')) continue
            var parsed = requirementName(candidate)
            if (parsed == null) {
                val egg = Regex("[#&]egg=([A-Za-z0-9][A-Za-z0-9._-]*)", RegexOption.IGNORE_CASE).find(candidate)
                if (egg != null) {
                    val name = egg.groupValues[1]
                    parsed = RequirementName(name, egg.range.last + 1 - name.length, name.length)
                }
            }
            if (parsed == null) continue
            val offset = logical.positions.getOrNull(candidateStart + parsed.start) ?: continue
            val range = SourceRange(offset, parsed.length)
            add(dependencyEntry(normalizePackageName(parsed.name), scope, range, listOf(range), parsed.name, range))
        }
    }

    private fun requirementsScope(path: String): String {
        val normalized = path.replace('\\', '/')
        val marker = "/requirements/"
        val index = normalized.lowercase().lastIndexOf(marker)
        return if (index >= 0) normalized.substring(index + 1) else normalized.substringAfterLast('/')
    }
}

File: src/core/python-manifest.ts
import { dependencyEntry } from './dependency';
import type { DependencyEntry, SourceRange } from './model';

interface TomlString {
  value: string;
  range: SourceRange;
  contentOffset: number;
  rawContent: string;
}

interface TomlAssignment {
  table: string[];
  key: string;
  keyRange: SourceRange;
  valueText: string;
  valueOffset: number;
}

export function normalizePythonPackageName(value: string): string {
  return value.trim().toLowerCase().replace(/[._-]+/g, '-');
}

function lineRecords(text: string): Array<{ text: string; offset: number }> {
  const out: Array<{ text: string; offset: number }> = [];
  let offset = 0;
  while (offset <= text.length) {
    const newline = text.indexOf('\n', offset);
    const rawEnd = newline < 0 ? text.length : newline;
    const end = rawEnd > offset && text[rawEnd - 1] === '\r' ? rawEnd - 1 : rawEnd;
    out.push({ text: text.slice(offset, end), offset });
    if (newline < 0) break;
    offset = newline + 1;
  }
  return out;
}

function commentStart(value: string): number {
  let quote: '"' | "'" | undefined;
  let escaped = false;
  for (let i = 0; i < value.length; i++) {
    const char = value[i]!;
    if (quote === '"' && escaped) escaped = false;
    else if (quote === '"' && char === '\\') escaped = true;
    else if (quote) {
      if (char === quote) quote = undefined;
    } else if (char === '"' || char === "'") quote = char;
    else if (char === '#') return i;
  }
  return value.length;
}

function splitDottedKey(raw: string): string[] | undefined {
  const parts: string[] = [];
  let start = 0;
  let quote: '"' | "'" | undefined;
  let escaped = false;
  const push = (end: number): boolean => {
    const token = raw.slice(start, end).trim();
    if (!token) return false;
    if ((token[0] === '"' || token[0] === "'") && token.at(-1) === token[0]) {
      parts.push(token.slice(1, -1));
    } else parts.push(token);
    return true;
  };
  for (let i = 0; i < raw.length; i++) {
    const char = raw[i]!;
    if (quote === '"' && escaped) escaped = false;
    else if (quote === '"' && char === '\\') escaped = true;
    else if (quote) {
      if (char === quote) quote = undefined;
    } else if (char === '"' || char === "'") quote = char;
    else if (char === '.') {
      if (!push(i)) return undefined;
      start = i + 1;
    }
  }
  return !quote && push(raw.length) ? parts : undefined;
}

function unquotedEquals(raw: string): number {
  let quote: '"' | "'" | undefined;
  let escaped = false;
  for (let i = 0; i < raw.length; i++) {
    const char = raw[i]!;
    if (quote === '"' && escaped) escaped = false;
    else if (quote === '"' && char === '\\') escaped = true;
    else if (quote) {
      if (char === quote) quote = undefined;
    } else if (char === '"' || char === "'") quote = char;
    else if (char === '=') return i;
  }
  return -1;
}

function nestingDelta(raw: string): number {
  let delta = 0;
  let quote: '"' | "'" | undefined;
  let escaped = false;
  for (const char of raw) {
    if (quote === '"' && escaped) escaped = false;
    else if (quote === '"' && char === '\\') escaped = true;
    else if (quote) {
      if (char === quote) quote = undefined;
    } else if (char === '"' || char === "'") quote = char;
    else if (char === '[' || char === '{') delta++;
    else if (char === ']' || char === '}') delta--;
  }
  return delta;
}

function tomlAssignments(text: string): TomlAssignment[] {
  const lines = lineRecords(text);
  const out: TomlAssignment[] = [];
  let table: string[] = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!;
    const content = line.text.slice(0, commentStart(line.text));
    const trimmed = content.trim();
    if (!trimmed) continue;
    if (trimmed.startsWith('[')) {
      if (!trimmed.startsWith('[[') && trimmed.endsWith(']')) {
        table = splitDottedKey(trimmed.slice(1, -1)) ?? [];
      } else table = [];
      continue;
    }
    const equals = unquotedEquals(content);
    if (equals < 0) continue;
    const rawKey = content.slice(0, equals);
    const keyToken = rawKey.trim();
    const keyParts = splitDottedKey(keyToken);
    if (!keyParts?.length) continue;
    const keyLeading = rawKey.length - rawKey.trimStart().length;
    const keyRawOffset = line.offset + keyLeading;
    const quoted = (keyToken[0] === '"' || keyToken[0] === "'") && keyToken.at(-1) === keyToken[0];
    const keyRange = {
      offset: keyRawOffset + (quoted ? 1 : 0),
      length: keyToken.length - (quoted ? 2 : 0),
    };
    let valueText = content.slice(equals + 1);
    const valueOffset = line.offset + equals + 1;
    let depth = nestingDelta(valueText);
    while (depth > 0 && i + 1 < lines.length) {
      const next = lines[++i]!;
      const nextContent = next.text.slice(0, commentStart(next.text));
      valueText += ' '.repeat(Math.max(0, next.offset - (valueOffset + valueText.length))) + nextContent;
      depth += nestingDelta(nextContent);
    }
    out.push({ table, key: keyParts.join('.'), keyRange, valueText, valueOffset });
  }
  return out;
}

function arrayStrings(assignment: TomlAssignment): TomlString[] {
  const out: TomlString[] = [];
  const raw = assignment.valueText;
  let inlineTableDepth = 0;
  for (let i = 0; i < raw.length; i++) {
    const quote = raw[i];
    if (quote === '{') {
      inlineTableDepth++;
      continue;
    }
    if (quote === '}') {
      inlineTableDepth = Math.max(0, inlineTableDepth - 1);
      continue;
    }
    if (quote !== '"' && quote !== "'") continue;
    if (raw.slice(i, i + 3) === quote.repeat(3)) {
      const end = raw.indexOf(quote.repeat(3), i + 3);
      if (end < 0) break;
      i = end + 2;
      continue;
    }
    const start = i;
    let escaped = false;
    for (i++; i < raw.length; i++) {
      const char = raw[i]!;
      if (quote === '"' && escaped) escaped = false;
      else if (quote === '"' && char === '\\') escaped = true;
      else if (char === quote) break;
    }
    if (i >= raw.length) break;
    const rawContent = raw.slice(start + 1, i);
    if (inlineTableDepth === 0) {
      out.push({
        value: rawContent,
        rawContent,
        contentOffset: assignment.valueOffset + start + 1,
        range: { offset: assignment.valueOffset + start, length: i - start + 1 },
      });
    }
  }
  return out;
}

function requirementName(raw: string): { name: string; start: number; length: number } | undefined {
  const match = /^\s*([A-Za-z0-9](?:[A-Za-z0-9._-]*[A-Za-z0-9])?)/.exec(raw);
  if (!match?.[1]) return undefined;
  const end = match[0].length;
  const next = raw[end];
  if (next && !/[\s[(<>=!~@;]/.test(next)) return undefined;
  return { name: match[1], start: match[0].length - match[1].length, length: match[1].length };
}

function entryFromRequirement(value: TomlString, scope: string): DependencyEntry | undefined {
  const parsed = requirementName(value.rawContent);
  if (!parsed) return undefined;
  const range = { offset: value.contentOffset + parsed.start, length: parsed.length };
  return dependencyEntry(normalizePythonPackageName(parsed.name), scope, range, [range], parsed.name, value.range);
}

function tableEquals(table: readonly string[], ...wanted: string[]): boolean {
  return table.length === wanted.length && table.every((part, index) => part === wanted[index]);
}

/** Static pyproject.toml parser. It records only dependency declarations and never evaluates a build backend. */
export function extractPyprojectDependencies(text: string): DependencyEntry[] {
  const out: DependencyEntry[] = [];
  for (const assignment of tomlAssignments(text)) {
    let arrayScope: string | undefined;
    if (tableEquals(assignment.table, 'project') && assignment.key === 'dependencies') arrayScope = 'project';
    else if (tableEquals(assignment.table, 'project', 'optional-dependencies')) arrayScope = `extra:${assignment.key}`;
    else if (tableEquals(assignment.table, 'dependency-groups')) arrayScope = `group:${assignment.key}`;
    else if (tableEquals(assignment.table, 'build-system') && assignment.key === 'requires') arrayScope = 'build-system';
    else if (tableEquals(assignment.table, 'tool', 'uv') && assignment.key === 'dev-dependencies') arrayScope = 'uv:dev';
    if (arrayScope) {
      for (const value of arrayStrings(assignment)) {
        const entry = entryFromRequirement(value, arrayScope);
        if (entry) out.push(entry);
      }
      continue;
    }

    let poetryScope: string | undefined;
    if (tableEquals(assignment.table, 'tool', 'poetry', 'dependencies')) poetryScope = 'poetry:main';
    else if (tableEquals(assignment.table, 'tool', 'poetry', 'dev-dependencies')) poetryScope = 'poetry:dev';
    else if (
      assignment.table.length === 5
      && assignment.table[0] === 'tool'
      && assignment.table[1] === 'poetry'
      && assignment.table[2] === 'group'
      && assignment.table[4] === 'dependencies'
    ) poetryScope = `poetry:group:${assignment.table[3]}`;
    if (poetryScope && assignment.key.toLowerCase() !== 'python') {
      const displayName = assignment.key;
      out.push(dependencyEntry(
        normalizePythonPackageName(displayName),
        poetryScope,
        assignment.keyRange,
        [assignment.keyRange],
        displayName,
        assignment.keyRange,
      ));
    }
  }
  return out;
}

function requirementsLogicalLines(text: string): Array<{ value: string; positions: number[] }> {
  const lines = lineRecords(text);
  const out: Array<{ value: string; positions: number[] }> = [];
  let value = '';
  let positions: number[] = [];
  for (const line of lines) {
    let part = line.text;
    const continued = /\\\s*$/.test(part);
    if (continued) part = part.replace(/\\\s*$/, '');
    for (let i = 0; i < part.length; i++) {
      value += part[i];
      positions.push(line.offset + i);
    }
    if (continued) {
      value += ' ';
      positions.push(line.offset + line.text.length);
    } else {
      out.push({ value, positions });
      value = '';
      positions = [];
    }
  }
  if (value) out.push({ value, positions });
  return out;
}

function requirementCommentStart(value: string): number {
  for (let i = 0; i < value.length; i++) {
    if (value[i] === '#' && (i === 0 || /\s/.test(value[i - 1]!))) return i;
  }
  return value.length;
}

/** Parses named requirements from one pip requirements file; includes and constraints remain separate manifests. */
export function extractRequirementsDependencies(text: string, scope = 'requirements.txt'): DependencyEntry[] {
  const out: DependencyEntry[] = [];
  for (const logical of requirementsLogicalLines(text)) {
    const withoutComment = logical.value.slice(0, requirementCommentStart(logical.value));
    const leading = withoutComment.length - withoutComment.trimStart().length;
    let candidate = withoutComment.trimStart();
    let candidateStart = leading;
    if (!candidate) continue;
    if (/^(?:-r|--requirement|-c|--constraint)(?:\s|=)/.test(candidate) || candidate.startsWith('--')) continue;
    const editable = /^(?:-e|--editable)(?:\s+|=)/.exec(candidate);
    if (editable) {
      candidateStart += editable[0].length;
      candidate = candidate.slice(editable[0].length);
    } else if (candidate.startsWith('-')) continue;

    let parsed = requirementName(candidate);
    if (!parsed) {
      const egg = /[#&]egg=([A-Za-z0-9][A-Za-z0-9._-]*)/i.exec(candidate);
      if (egg?.[1] && egg.index >= 0) {
        parsed = { name: egg[1], start: egg.index + egg[0].length - egg[1].length, length: egg[1].length };
      }
    }
    if (!parsed) continue;
    const logicalStart = candidateStart + parsed.start;
    const offset = logical.positions[logicalStart];
    if (offset === undefined) continue;
    const range = { offset, length: parsed.length };
    out.push(dependencyEntry(
      normalizePythonPackageName(parsed.name),
      scope,
      range,
      [range],
      parsed.name,
      range,
    ));
  }
  return out;
}

export function pythonRequirementsScope(path: string): string {
  const normalized = path.replace(/\\/g, '/');
  const marker = '/requirements/';
  const index = normalized.toLowerCase().lastIndexOf(marker);
  if (index >= 0) return normalized.slice(index + 1);
  return normalized.split('/').at(-1) ?? 'requirements.txt';
}
